### PR TITLE
Add experimental format-preserving mode to `@babel/generator`

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | https://github.com/babel/website/pull/2994
| Any Dependency Changes?  |
| License                  | MIT

This PR implements a `experimental_preserveFormat: true` option for `@babel/generator` that lets it print code fully respecting its original format.

This is done by parsing the code with `tokens: true` and by passing the original source code to `@babel/generator`: whenever it needs to print a token, it tries to match it with one of the tokens in the input file and prints it in the same location, if possible. When there are new tokens (for new nodes), it prints them as compact as possible so that the next token has a chance of being printed in the correct position, rather than being "pushed ahead".

This option currently only works with `retainLines: true`. If throws an error if you don't explicitly set it, because in the future I plan to investigate removing that restriction (so that, for example, a codemod could inject new nodes in the middle and the rest gets shifted without being re-formatted).

This mode works for JS (including proposals), JSX and TypeScript. Unfortunately our Flow AST makes it too difficult to apply the same strategy there.

`packages/babel-generator/test/preserve-format.js` is verifying that we are able to parse and re-print all the existing generator tests and get exactly the same output. There is a list of failures -- I divided them in a few groups, and they are all expected failures because they are testing some cases in which the options explicitly ask to re-print the code in a different way.

A lot of changes have been extracted to separate PRs, at https://github.com/babel/babel/milestone/61. There are still a couple PRs that this one depends on (I currently just cherry-picked them), so go review them first! :)

### CODE REVIEW

File: packages/babel-generator/test/fixtures/sourcemaps/real-world-babel-file1/source-map.json [hunk 1]
```diff
@@ -74,6 +74,6 @@
...
...sB,IAAI,CAACrB,GAAG,CAACiB,WAAW,CAAC;EACvB;EAEAvC,UAAU,CAAC8C,GAAG,CAACP,WAAW,EAAEC,eAAe,CAAC;AAC9C;AAEA,SAASK,oBAAoBA,CAAClB,QAAQ,EAAE,MAAM,EAAEN,GAAG,EAAE,MAAM,EAAE;EAC3D,MAAMsB,IAAI,GAAGxC,UAAU,CAACiC,GAAG,CAACf,GAAG,CAAC,IAAInB,GAAG,CAAC,MAAM,CAAC;EAC/CyC,IAAI,CAACI,MAAM,CAACpB,QAAQ,CAAC;EAErB,IAAIgB,IAAI,CAACK,IAAI,KAAK,CAAC,EAAE;IACnB7C,UAAU,CAAC4C,MAAM,CAAC1B,GAAG,CAAC;IAEtB,IAAI,CAACd,UAAU,EAAE;MACfD,UAAU,CAACyC,MAAM,CAAC1B,GAAG,CAAC;IACxB,CAAC,MAAM;MACLhB,OAAO,CAAC4C,OAAO,CAAC5B,GAAG,CAAC;IACtB;EACF;AACF;AAEA,SAASF,WAAWA,CAACQ,QAAQ,EAAE,MAAM,EAAE;EACrC,MAAMgB,IAAI,GAAG3C,UAAU,CAACoC,GAAG,CAACT,QAAQ,CAAC;EACrC,IAAI,CAACgB,IAAI,EAAE;EAEX,KAAK,MAAMtB,GAAG,IAAIsB,IAAI,EAAE;IACtBE,oBAAoB,CAAClB,QAAQ,EAAEN,GAAG,CAAC;EACrC;EACArB,UAAU,CAAC+C,MAAM,CAACpB,QAAQ,CAAC;AAC7B;AAEA,SAASjB,eAAeA,CAAA,CAAE,EAAE,GAAG,CAAC;EAC9B,MAAMwC,OAAO,GAAGtD,aAAa,CAACuD,MAAM,CAACC,IAAI,CAACC,GAAG,CAAC;EAE9C,IAAI;IACF,OAAOC,OAAO,CAACC,GAAG,CAACC,gBAAgB,GAC/BN,OAAO,CAAC,UAAU,CAAC,GACnBO,QAAQ,CAACH,OAAO,CAACI,QAAQ,CAACC,IAAI,CAAC,IAAI,CAAC,GACpCT,OAAO,CAAC,UAAU,CAAC,GACnBA,OAAO,CAAC,4BAA4B,CAAC;EAC3C,CAAC,CAAC,OAAOU,GAAG,EAAE;IACZpC,OAAO,CAACqC,KAAK,CACX,yEAAyE,GACvE,6DACJ,CAAC;IACD,MAAMD,GAAG;EACX;AACF",
```
Comment (nicolo-ribaudo) [inline packages/babel-generator/test/fixtures/sourcemaps/real-world-babel-file1/source-map.json:77]: The difference here is that a KAAK is replaced by EAAE,GAAG. KAAK means [+5, +0, +0, +0] and EAAE,GAAG means [+2,0,0,+2], [+3,0,0,+3], so the result is exactly the same.

**Comment by nicolo-ribaudo** [comment]
@JLHwung @liuxingbaoyu This is ready for review! The main logic is in the new `findToken`&co functions in `babel-generator/src/printer.ts`, and the rest of the files in the `generator/` folder have some extra logic for those cases where the "automatic" logic is not smart enough.

There are still 3 PRs it depends on at https://github.com/babel/babel/milestone/61, for now I have just cherry-picked them.

**Comment by nicolo-ribaudo** [comment]
Mmm this PR currently makes the non-format-preserving 2x slow, I need to figure out why :thinking:

**Comment by nicolo-ribaudo** [comment]
All the dependency PRs have been merged :)

**Review by liuxingbaoyu — COMMENTED** [review]
It would be nice if we could somehow detect whether createParenthesizedExpressions was enabled correctly.

**Comment by nicolo-ribaudo** [comment]
To try this PR in a different repo:
- Go in the PR "Checks", in the "E2E tests" workflow (_not_ the breaking one)
- Download the "verdaccio-workspace" artifact
- It contains tgz archives for all the packages. You only need the one for `@babel/generator`
- Tell Yarn (or any other package manager) to use that archive, with `"resolutions": { "@babel/generator": "./generator-7.25.7.tgz" }`. You might also need to add `"resolutions": { "@babel/parser": "^7.25.0" }`
- Enable the `preserveFormat: true` options of `@babel/generator`, and pass the original source code as the third parameter. Enable the `tokens: true, createParenthesizedExpression: true` options of `@babel/parser`.

**Comment by nicolo-ribaudo** [comment]
> It would be nice if we could somehow detect whether createParenthesizedExpressions was enabled correctly.

Maybe if we see `extra.parenthesized` somewhere we can error?

**Comment by liuxingbaoyu** [comment]
I added a rudimentary test in `packages\babel-generator\test\preserve-format.js` and it seems that currently there is some broken output for generated code without position information.
I am a bit worried that this is more likely to happen in practice, maybe we can keep a space for code without position information.
```js
import { parse } from "@babel/parser";
import { cloneNode } from "@babel/types";
import path from "path";
import fixtures from "@babel/helper-fixtures";
import * as babel from "@babel/core";
import pluginTransformTypeScript from "@babel/plugin-transform-typescript";
import { commonJS } from "$repo-utils";

import _generate from "../lib/index.js";
const generate = _generate.default || _generate;

const { __dirname } = commonJS(import.meta.url);

const suites = (fixtures.default || fixtures)(path.join(__dirname, "fixtures"));

describe("preserveFormat", () => {
  describe("generation", () => {
    suites.forEach(testSuite => {
      describe(`${testSuite.title}`, () => {
        testSuite.tests.forEach(function (task) {
          if (task.options.throws) return;

          const testFn = task.disabled ? it.skip : it;

          testFn(task.title, () => {
            const input = task.actual.code.replace(/\r\n?/g, "\n");
            const parserOpts = {
              filename: task.actual.loc,
              plugins: task.options.plugins || [],
              strictMode: task.options.strictMode === false ? false : true,
              sourceType: "module",
              ...task.options.parserOpts,
              createParenthesizedExpressions: true,
              tokens: true,
            };
            const ast = parse(input, parserOpts);
            const options = {
              sourceFileName: path.relative(__dirname, task.actual.loc),
              ...task.options,
              retainLines: true,
              preserveFormat: true,
              comments: true,
              jsescOption: null,
              minified: false,
              compact: false,
            };

            const ok = generate(ast, options, input).code === input;
            const shouldFail = FAILURES.some(f =>
              task.actual.loc.replace(/\\/g, "/").endsWith(f),
            );

            if (!ok && shouldFail) {
              expect(1).toBe(1);
              return;
            }
            if (ok) {
              expect(shouldFail).toBe(false);
            }

            expect(generate(ast, options, input).code).toBe(input);
          });

          testFn(`${task.title} without loc`, () => {
            const input = task.actual.code.replace(/\r\n?/g, "\n");

            const parserOpts = {
              filename: task.actual.loc,
              plugins: task.options.plugins || [],
              strictMode: task.options.strictMode === false ? false : true,
              sourceType: "module",
              ...task.options.parserOpts,
              createParenthesizedExpressions: true,
              tokens: true,
            };
            const ast = parse(input, parserOpts);
            const options = {
              sourceFileName: path.relative(__dirname, task.actual.loc),
              ...task.options,
              retainLines: true,
              preserveFormat: true,
              comments: true,
              jsescOption: null,
              minified: false,
              compact: false,
            };

            ast.program = cloneNode(ast.program, true, true);

            const code = generate(ast, options, input).code;
            try {
              parse(code, parserOpts);
            } catch (error) {
              expect(code).toBe(input);
            }
          });
        });
      });
    });
  });

  describe("code transforms", () => {
    it("type stripping with arrow functions", () => {
      const input = `
        const a = (x: string):
                        number => 2 as const;
      `;
      const expected = `
        const a = (x        )=>
                                  2         ;
      `;

      const out = babel.transformSync(input, {
        configFile: false,
        plugins: [pluginTransformTypeScript],
        parserOpts: {
          createParenthesizedExpressions: true,
          tokens: true,
        },
        generatorOpts: {
          retainLines: true,
          preserveFormat: true,
        },
      });

      expect(out.code.trimEnd()).toBe(expected.trimEnd());
    });

    it("identifier renaming", () => {
      const input = `
        const foo = 3;
        const bar = 3;

        foo( x, y, z );
        bar( x, y, z );
      `;
      const expected = `
        const x   = 3;
        const longer=3;

        x  ( x, y, z );
        longer(x,y,z );
      `;

      const out = babel.transformSync(input, {
        configFile: false,
        plugins: [
          () => ({
            visitor: {
              Program(path) {
                path.scope.rename("foo", "x");
                path.scope.rename("bar", "longer");
              },
            },
          }),
        ],
        parserOpts: {
          createParenthesizedExpressions: true,
          tokens: true,
        },
        generatorOpts: {
          retainLines: true,
          preserveFormat: true,
        },
      });

      expect(out.code.trimEnd()).toBe(expected.trimEnd());
    });
  });
});

const FAILURES = [
  // These tests are either explicitly about re-formatting the decorators position,
  // or about an old decorators version
  "comments/decorators-after-export-to-before/input.js",
  "comments/decorators-before-export-to-after/input.js",
  "decorators/decorator-call-expression/input.js",
  "decorators/decorator-parenthesized-expression/input.js",
  "decorators/decorator-parenthesized-expression-createParenthesizedExpression/input.js",
  "decoratorsBeforeExport/false-to-true/input.js",
  "decoratorsBeforeExport/true-to-false/input.js",

  // The 'preserveFormat' option does not fully support Flow
  "flow/array-types/input.js",
  "flow/arrow-functions/input.js",
  "flow/call-properties/input.js",
  "flow/declare-module/input.js",
  "flow/declare-exports/input.js",
  "flow/declare-statements/input.js",
  "flow/indexed-access-types/input.js",
  "flow/interfaces-module-and-script/input.js",
  "flow/iterator-inside-declare/input.js",
  "flow/iterator-inside-interface/input.js",
  "flow/iterator-inside-types/input.js",
  "flow/object-literal-types/input.js",
  "flow/opaque-type-alias/input.js",
  "flow/parantheses/input.js",
  "flow/this-param/input.js",
  "flow/tuples/input.js",
  "flow/type-alias/input.js",
  "flow/type-annotations/input.js",
  "flow/type-union-intersection/input.js",
  "flow/typecasts/input.js",
  "flowUsesCommas/ObjectExpression/input.js",

  // These tests are explicitly about changing the output
  "importAttributesKeyword/assertions-assert-to-with/input.js",
  "importAttributesKeyword/assertions-assert-to-with-legacy/input.js",
  "importAttributesKeyword/assertions-with-to-assert/input.js",
  "importAttributesKeyword/assertions-with-to-default/input.js",
  "importAttributesKeyword/assertions-with-to-with-legacy/input.js",
  "importAttributesKeyword/attributes-assert-to-default/input.js",
  "importAttributesKeyword/attributes-assert-to-with/input.js",
  "importAttributesKeyword/attributes-assert-to-with-legacy/input.js",
  "importAttributesKeyword/attributes-with-to-assert/input.js",
  "importAttributesKeyword/attributes-with-to-default/input.js",
  "importAttributesKeyword/attributes-with-to-with-legacy/input.js",
  "importAttributesKeyword/legacy-module-attributes-to-assert/input.js",
  "importAttributesKeyword/legacy-module-attributes-to-with/input.js",

  // These tests are about old proposals
  "types/Decorator/input.js",

  // We explicitly always print `module X {}` as `namespace X {}`
  "typescript/module-namespace-head/input.js",
  "typescript/module-namespace-head-declare/input.js",
  "typescript/module-namespace-head-export/input.js",
  "typescript/export-declare/input.js",
];

```

**Comment by nicolo-ribaudo** [comment]
@liuxingbaoyu Only one of those failures is an actual bug, fixed by https://github.com/babel/babel/pull/16788. The other are either:
- for Flow, which `preserveFormat` does not support
- for cases in which the generator "transpiles" code (for example, when it's configured to parse import assertions and emit import attributes) and thus it's missing the correct plugin to re-parse

You can see the actual error by replacing the assertion with `expect([code, error]).toBe([input, null]);`.

The mode without spaces is already used and well-tested, because it's the same logic that `compact: true` uses.

File: packages/babel-generator/src/generators/types.ts [hunk 2]
```diff
@@ -3,9 +3,27 @@ import { isAssignmentPattern, isIdentifier } from "@babel/types";
...
 import jsesc from "jsesc";
 
+let lastRawIdentNode: t.Identifier | null = null;
+let lastRawIdentResult: string = "";
+export function _getRawIdentifier(this: Printer, node: t.Identifier) {
+  if (node === lastRawIdentNode) return lastRawIdentResult;
+  lastRawIdentNode = node;
+
+  const { name } = node;
+  const token = this._findToken(tok => tok.value === name);
```
Comment (liuxingbaoyu) [inline packages/babel-generator/src/generators/types.ts:13]: Maybe we should also compare positions? If there are multiple tokens with the same value.

**Comment by liuxingbaoyu** [comment]
Non-blocking: It seems that the benchmark (real-case) is 10% slower, I investigated and unfortunately did not find the reason.(Even if I replace all `this.format.preserveFormat` to `false`)

**Comment by nicolo-ribaudo** [comment]
Oh, I'll take a look

**Comment by liuxingbaoyu** [comment]
This fixes the performance regression.
It seems the problem comes from closures, which is weird.

Thanks a lot for your commits, I used `git bisect` to finally find it.

```
diff --git a/packages/babel-generator/src/node/parentheses.ts b/packages/babel-generator/src/node/parentheses.ts
index 63db36c3ab..c9419bfe39 100644
--- a/packages/babel-generator/src/node/parentheses.ts
+++ b/packages/babel-generator/src/node/parentheses.ts
@@ -463,7 +463,11 @@ export function Identifier(
     }
   }
 
-  if (getRawIdentifier(node) !== node.name) {
+  const name =
+    typeof getRawIdentifier === "function"
+      ? getRawIdentifier(node)
+      : getRawIdentifier.name;
+  if (name !== node.name) {
     return false;
   }
 
diff --git a/packages/babel-generator/src/printer.ts b/packages/babel-generator/src/printer.ts
index 75f2e7e663..ce9c8acd61 100644
--- a/packages/babel-generator/src/printer.ts
+++ b/packages/babel-generator/src/printer.ts
@@ -879,9 +879,7 @@ class Printer {
         parent,
         this.tokenContext,
         this.inForStatementInit,
-        format.preserveFormat
-          ? node => this._getRawIdentifier(node)
-          : node => node.name,
+        format.preserveFormat ? this._getRawIdentifier.bind(this, node) : node,
       );
 
     if (

```

**Comment by nicolo-ribaudo** [comment]
@liuxingbaoyu Perf regression fixed. I also added https://github.com/babel/babel/pull/16708#issuecomment-2321260625 to the test suite.

**Comment by nicolo-ribaudo** [comment]
@JLHwung I'm thinking of renaming this back to `experimental_preserveFormat`, until when the mode without `retainLines: true` (which can go in a separate release) is ready.

**Comment by kriskowal** [comment]
@nicolo-ribaudo Thank you. I’ve begun some integration. My first stab (following your instructions as closely as my crossed-eyes permit) has not worked yet, but I’ll review this PR to see if I’ve missed anything. If you with to follow along, the changes are in endojs/endo PR 2444 and the spot-check test fixtures are linked below. The expectation is that mangling.js should look like unmangled.js after an ESM to Program transformation.

https://github.com/endojs/endo/pull/2444/files#diff-51ac3e3c5313c6e06d439478e7b3a7a5f96f16a41eab6006a2e3ce213e8a6790

**Comment by kriskowal** [comment]
@nicolo-ribaudo I’m happy to report my second attempt to integrate is working end-to-end for a spot-check test. https://github.com/endojs/endo/pull/2444

File: packages/babel-generator/src/generators/types.ts [hunk 3]
```diff
@@ -96,14 +118,16 @@ export function ArrayExpression(this: Printer, node: t.ArrayExpression) {
     if (elem) {
       if (i > 0) this.space();
       this.print(elem);
-      if (i < len - 1) this.token(",");
+      if (i < len - 1 || this.shouldPrintTrailingComma("]")) {
+        this.token(",", false, i);
```
Comment (JLHwung) [inline packages/babel-generator/src/generators/types.ts:122]: Q1: For each comma, we initiate a linear token scan from the start of the array expression, doesn't this imply an O(#element^2) search over the tokens array? Can we somehow reuse the token index of the previously printed comma?

Q2: Should this change be also cherry-picked to the `TupleExpression` generator?

**Review by JLHwung — COMMENTED** [review]
Awesome work. Left a few comments. It would be great if you can add some JSDoc comments to the newly introduced printer methods.

**Comment by nicolo-ribaudo** [comment]
@JLHwung Done! While trying to figure out how to avoid the `O(n^2)` case, I changed the "find tokens corresponding to each node" logic to happen eagerly rather than lazily (since in practice we would do it for every node anyway), and that gave a 2x speedup when trying to print `@babel/standalone`'s unminified bundle under this new mode (1s rather than 2s).

**Review by liuxingbaoyu — APPROVED** [review]
Incredibly amazing, thanks!
We can use `if (this._tokenMap)` instead of `if (!this.format.preserveFormat)`, which is smaller and faster. :)

**Comment by nicolo-ribaudo** [comment]
@liuxingbaoyu Done. I left `.format.preserveFormat` in a few places where we don't actually use the tokens, since it would have otherwise been confusing.

Thanks both for the reviews!

**Pull request merged**